File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/SourceEnum.java
package ai.ragu.charactermanager.enumeration;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.HashMap;
import java.util.Map;

public enum SourceEnum implements WeaponTraitEnum {
    CORE_RULEBOOK("coreRulebook"),
    LOST_OMENS("lostOmensCharacterGuide");

    private final String value;
    private final static Map<String, SourceEnum> CONSTANTS = new HashMap<>();

    static {
        for (SourceEnum c: values()) {
            CONSTANTS.put(c.value, c);
        }
    }

    SourceEnum(String value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return this.value;
    }

    @JsonValue
    public String value() {
        return this.value;
    }

    @JsonCreator
    public static SourceEnum fromValue(String value) {
        SourceEnum constant = CONSTANTS.get(value);
        if (constant == null) {
            throw new IllegalArgumentException(value);
        } else {
            return constant;
        }
    }
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/SourceEnum.kt
package ai.ragu.charactermanager.enumeration

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import java.util.*

enum class SourceEnum(private val value: String) : WeaponTraitEnum {
    CORE_RULEBOOK("coreRulebook"), LOST_OMENS("lostOmensCharacterGuide");

    companion object {
        private val CONSTANTS: MutableMap<String, SourceEnum> = HashMap()
        @JsonCreator
        fun fromValue(value: String?): SourceEnum {
            val constant = CONSTANTS[value]
            return constant ?: throw IllegalArgumentException(value)
        }

        init {
            for (c in values()) {
                CONSTANTS[ai.ragu.charactermanager.enumeration.c.value] = ai.ragu.charactermanager.enumeration.c
            }
        }
    }

    override fun toString(): String {
        return value
    }

    @JsonValue
    fun value(): String {
        return value
    }

}